### DISCUSSION TITLE
should `useTsNode` even be set in the config?

### DISCUSSION THREAD

**Discussion by jgburet** [mikro-orm/mikro-orm#1340] [Q&A] [answered]
I'm not quite sure I understand what is the way to go between `useTsNode` and `tsNode` :/

This is what I have right now : 
```json
# package.json

    "scripts": {
        "build": "rm -rf ./build; yarn; tsc --build; yarn --production"
    },
    "mikro-orm": {
        "useTsNode": false,
        "tsConfigPath": "./tsconfig.json",
        "configPaths": [
            "./src/mikro-orm.config.ts",
            "./build/src/mikro-orm.config.js"
        ]
    },
```

```typescript
// src/mikro-orm.config.ts

    tsNode: process.env.NODE_ENV !== 'production',
    entities: ['./build/src/**/*.entity.js'],
    entitiesTs: ['./src/**/*.entity.ts'],
```

During development, if I run `$ yarn mikro-orm migration:up` I get this error message:
```
Error: MikroORM config file not found in ['./build/src/mikro-orm.config.js', './mikro-orm.config.js']
```
The CLI didn't understand it had to look at the config with a .ts extension.
From here I'm pretty sure you are going to tell me to the the option to `true` in the `package.json` 🙃 

Sure, let's try that:
```json
# package.json

    "mikro-orm": {
        "useTsNode": true,
```

`$ yarn build` , yep, shipping it to prod
`$ yarn mikro-orm migration:up`
```yarn run v1.22.4
$ /Users/jgburet/Development/welii/backend/node_modules/.bin/mikro-orm migration:up
(node:35013) UnhandledPromiseRejectionWarning: Error: Cannot find module 'ts-node'
```

What path do you recommend from here?

**Comment by B4nan** [comment]
For starters, `useTsNode` is not an ORM configuration at all, that is only a setting for the CLI (only available via `package.json`) - if you want the CLI to understand TS files (like TS migrations or TS CLI config), you should enable that. You need to have ts-node in your dependencies - and if you want to use the CLI in production, it has to be `dependencies`, not just `devDependencies`. 

On the other hand, `tsNode` is - but it is something to use only when testing to enforce discovery on TS files, and you should not need to use it unless the automatic ts-node detection fails. 

I'd suggest to look into the example projets, they should be all set up correctly in terms of CLI.

https://github.com/mikro-orm/mikro-orm-examples

> The CLI didn't understand it had to look at the config with a .ts extension.

Nope, the CLI ignores them becausse you told it to do so by setting `useTsNode` to false (which is the default btw).

**Comment by jgburet** [comment]
Thank you for the really quick answer!

> and if you want to use the CLI in production, it has to be dependencies, not just devDependencies.

Well, I don't see how else I could run migrations on the DB if I don't have it in production. Or did I miss something?
I think everything links to this very point here :/
So do you suggest to keep the migrations to plain JS files?

Since I'm compiling the TS into JS, I shouldn't have to set `ts-node` as a direct dependancy, and only keep it as a dev dependancy, just as it has been done on the examples you linked https://github.com/mikro-orm/koa-ts-example-app/blob/master/package.json#L27-L30
However there's something I'd like to point at in this example repo: useTsNode is set to `true`. So how do you do if dev dependancies are cleaned?

**Comment by B4nan** [comment]
> Well, I don't see how else I could run migrations on the DB if I don't have it in production. Or did I miss something?

You don't need CLI to run migrations, you can do so programatically (e.g. on app start). That is usually much better suited than doing it manually, the CLI is more of a dev tool (at least to me personally). 

> Since I'm compiling the TS into JS, I shouldn't have to set ts-node as a direct depency, and only keep it as a dev dependancy

If you want to execute TS migration via the ORM CLI, you need it as direct dependency. Of course you can compile them and run the JS migrations in production - but that requires some changes in the migrations config, as by default it will look for TS files. Namely you need to adjust `pattern` and possibly also `path`.

https://mikro-orm.io/docs/migrations/#configuration

> just as it has been done on the examples you linked https://github.com/mikro-orm/koa-ts-example-app/blob/master/package.json#L27-L30

Yes, but you can also see that the CLI is a dev dependency there. You are talking about using the CLI in production, so the CLI package would have to direct dependency too.

**Comment by B4nan** [comment]
Same applies to the ORM config, you can compile it and the ORM will pick the JS one if it:

1. exists on the configured path (you can verify that via the `debug` cli command locally)
2. ts-node is not detected or enforced via `tsNode: true`

That way the configs with `.ts` extension will be ignored and the first available `.js` config will be used:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/ConfigurationLoader.ts#L58

**Comment by jgburet** [comment]
Very clear regarding the CLI :)

So if I understand correctly, I should keep the `tsNode: NODE_ENV !== prod` in my mikro-orm.config.ts in case the developer has something in his ./build directory for some reason while developing.

**Comment by B4nan** [comment, marked as answer]
I'd say remove it, and let the autodetection do its job. It depends on your use case, but personally we were using `NODE_ENV === 'development'` in the dev server, not just locally, so there we also wanted to use "production settings" (compiled files). You should need to set that only when running tests (when running via ts-node), when the autodetection fails.

https://github.com/mikro-orm/mikro-orm/blob/v4.4.0/packages/core/src/utils/Utils.ts#L519

All what the `tsNode` flag does is to force usage of `entitiesTs` instead of `entities`. It is needed because we need to discover the right files - you could have the project compiled, and then there would be two copies of each class, one TS and one JS. It is needed only for the folder based discocovery, where we need to require the right file (TS or JS one). 

If you would use class references instead, e.g. `entities: [Author, Book, ...]`, then none of this would be needed, as the reference would already lead to the right file (when running via ts-node, it would be the TS file, and vice versa).

**Comment by jgburet** [comment]
I did just like you said:
- set `useTsNode: true` in package.json
- removed `tsNode: qsdqsd` from mikro-orm.config.ts
- stopped using the CLI on production env
- set the migration to run automatically on startup (I defined a RUN_MIGRATIONS_ON_STARTUP env var in case I need to skip that, though)

and it just works :)
The third point was really what was breaking it all.

However, I have to admit I didn't feel very comfortable not having a `up` and (especially) a `down` script for the migrations, but I realized that anyway when there's a `down` to run in production (sh*t sometimes happen) we do it from our terminal, not using any CI/CD action ^^

Thank you for your help :)
